### PR TITLE
bpo-11233: Create availability directive for documentation

### PR DESCRIPTION
**Pull request by csabella**
Original patch by Georg Brandl.

https://bugs.python.org/issue11233

### CODE REVIEW

**Review by vstinner — CHANGES_REQUESTED** [review]
I like the general approach, but there are a few issues:

* "make html" generates 3 warnings

/home/vstinner/prog/python/master/Doc/library/os.rst:1111: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/vstinner/prog/python/master/Doc/library/os.rst:1179: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/vstinner/prog/python/master/Doc/library/os.rst:3555: WARNING: Explicit markup ends without a blank line; unexpected unindent.

* "Availability: Unix." is used outside os.rst: asyncio, signal, time, etc. documentation.

* Because the availability is used in multiple files, maybe the definition of availability should be described somewhere else? Sorry I have no better proposal, so maybe allos.rst is fine.

**Comment by csabella** [comment]
@vstinner   
* I fixed the 'make html' warnings.  Sorry that I missed those the first time.  
* Based on the comment about other sources, I also added the availability directive to the other sources.  I did it in its own commit so it could be reversed if it's too much.
* As far as where to store the definition of availability, I looked for a global 'notes'-type page to add it to, but didn't find one.  The closest concept that I could find is maybe 'CPython Implementation Detail', but that doesn't have a note defining it.  For now, I've moved it to the library intro page.  It doesn't flow well with the opening paragraphs, so maybe it should be its own page at the end of the TOC?

File: Doc/library/allos.rst [hunk 1]
```diff
@@ -27,3 +27,4 @@ but they are available on most other systems as well.  Here's an overview:
    platform.rst
    errno.rst
    ctypes.rst
+
```
Comment (vstinner) [inline Doc/library/allos.rst]: Please revert this change.

File: Doc/tools/extensions/pyspecific.py [hunk 2]
```diff
@@ -131,6 +131,26 @@ def run(self):
...
+    final_argument_whitespace = True
+
+    def run(self):
+        pnode = nodes.paragraph(classes=['availability'])
+        n, m = self.state.inline_text(':ref:`Availability <availability>`: ',
+                                      self.lineno)
+        pnode.extend(n + m)
+        n, m = self.state.inline_text(self.arguments[0], self.lineno)
+        pnode.extend(n + m)
+        pnode += nodes.Text('.', '.')
```
Comment (vstinner) [inline Doc/tools/extensions/pyspecific.py]: I dislike the idea of adding a trailing dot in the macro. Can you please remove it? I prefer to give most freedom to the user of the macro.

**Review by vstinner — COMMENTED** [review]
Hum, I would prefer that the macro doesn't add a trailing dot. I prefer to give more freedom to the user of the macro to decide how to render the text.

File: Doc/tools/extensions/pyspecific.py (hunk 2 above)
Comment (csabella) [reply Doc/tools/extensions/pyspecific.py]: Removed.

File: Doc/library/allos.rst (hunk 1 above)
Comment (csabella) [reply Doc/library/allos.rst]: Done.

**Comment by vstinner** [comment]
Thanks @csabella! I checked the full PR and it seems like all "Availability" lines now ends with a dot :-) It's more consistent. For the exact location of the definition of Availability, if someone complains, we can now easily change it, since it has a reST label. The whole PR LGTM, I merged it.

I don't want it to be backported. It seems too intrusive to backport it.

**Comment by 1st1** [comment]
a post-commit LGTM from me.

**Comment by 1st1** [comment]
BTW, I think this PR should be backported to 3.7, otherwise we might have unnecessary conflicts when backporting docs updates to 3.7.

**Comment by vstinner** [comment]
> BTW, I think this PR should be backported to 3.7, otherwise we might have unnecessary conflicts when backporting docs updates to 3.7.

If you want to avoid conflicts, in that case, should we also apply it to 3.6 and 2.7?

**Comment by csabella** [comment]
Do you want me to cherry pick to 3.7 or is Miss Islington able to do it post-merge?

**Comment by 1st1** [comment]
> Do you want me to cherry pick to 3.7 or is Miss Islington able to do it post-merge?

Apparently she can't :(  Could you please open a PR for 3.7 branch?
